### ISSUE TITLE
Link checker

### ISSUE THREAD

**Issue by jesusff** [cf-convention/cf-convention.github.io#285]
The automatic link checker reports problems such as:
```
[ERR] file:///github/workspace/discussion.html | Cannot find file
```
which are the result of relative links (i.e. just discussion.html) that work when rendered in the website, but not in the .md as seen by the action script. We could either fix the checker to skip those or just use absolute links (i.e. https://cfconventions.org/discussion.html) in all markdown files.

As is, there will be a report with broken links every week and after each commit, whether there are really broken links or not.

**Comment by DocOtak** [comment]
Assuming we can sort out the issue with relative links, I think we should modify the link checker action to add a comment on the specific pull request rather than make a new issue when it finds problems (in PRs at least).

**Comment by DocOtak** [comment]
Also that comment should be updated when the PR changes: perhaps using something like https://github.com/marketplace/actions/sticky-pull-request-comment

**Comment by cofinoa** [comment]
This relates to #181

**Comment by cofinoa** [comment]
We need a tidy-up how to link self-pages on jekyll using `liquid` filters, `relative_url` and `absolute_url`: https://jekyllrb.com/docs/liquid/filters/

probably a recommendation on not use absolute url, to refer cf-convention page, should be done and corresponding check on PR should be created.

**Comment by ethanrd** [comment]
PR #298 addresses the ".md" vs ".html" links. (And some broken Reference Links.)

**Comment by cofinoa** [comment]
I have been running locally the link checker (lychee) and there is plenty link errors that will be hard to handle. 

@ethanrd instead of running the link checker over the repo itself with the no-rendered .md it may be would better to made over the rendered `_site` directory after the github action which runs jekyll and rended all .md to .html.

Some filtering needs to be done to no-check some paths, take a look to the attach report: 
[lychee_out_local.md](https://github.com/cf-convention/cf-convention.github.io/files/9668015/lychee_out_local.md)

where I have run `lychee` over the jekyll rendered `_site` directory excluding the `_site_/Data` directory which has many legacy broken links, and non UTF-8/ASCII HTML files which raise an error on running `lychee`

**Comment by cofinoa** [comment]
I have created my own fork to test the PR actions, and here you can see a partial report, where it has stopped because one of the files has invalid UTF8 encoding:
https://github.com/cofinoa/cf-convention.github.io/actions/runs/3143322116/jobs/5107923254#step:3:4332

**Comment by cofinoa** [comment]
@ethanrd I think linking to .md instead .html it's the way to linking internal pages rendered by Jekyll. 

More info:
https://mademistakes.com/mastering-jekyll/how-to-link/#jekyll-relative-links-plugin

**Comment by ethanrd** [comment]
Looks like (in the [latest CF website build](https://github.com/cf-convention/cf-convention.github.io/actions/runs/3145089727/jobs/5111937582)) the automatic GitHub Pages build uses the [jekyll-relative-links](https://github.com/benbalter/jekyll-relative-links) plugin which converts relative ".md" links to ".html" in the resulting web pages. So I think we are OK using either ".md" or ".html".

I think it would be good to be consistent. I was thinking we should go with ".md" since the link checker succeeds on those and not ".html". But, as you say, we could switch to checking the `_site` resulting from out test build.

**Comment by ethanrd** [comment]
If using ".md" links fixes the link checker and produces ".html" links in the resulting HTML pages, I would prefer keeping the link checker independent of the Jekyll build (just for "separation of concerns", I guess).

On the other hand, I think the scheduled link checking should actually be run against the https://cfconventions.org/ site. That would cover the other repos that contribute to the web site, e.g., `mailing-list-archive` and `cf-training`.

**Comment by ethanrd** [comment]
I just pushed another commit to PR #298. The only remaining [failing links](https://github.com/cf-convention/cf-convention.github.io/actions/runs/3155837203/jobs/5134925245) are

- A link to the `CODE_OF_CONDUCT.md` [file in GH](https:/github.com/cf-convention/cf-conventions/blob/master/CODE_OF_CONDUCT.md) that the link checker appears to handle like a local file, I guess because it is an ".md" file. (I think we'll have to configure this one to be ignored.)
- Some links to the old cf-metadata email list archive were fixed by switching from "http:" to "https:". A number of links to specific posts are still broken, I believe due to the archive being reorganized/renumbered some years back. (We can try to figure out what posts the links are intended to reference. Should we also convert to point to the new email list archive?)
- Several links to the CF Editor which is the correct URL but a bit slow and the link checker times out. (We may be able to configure the default timeout the link checker uses.)

**Comment by cofinoa** [comment]
@ethanrd , I get lost here:
>On the other hand, I think the scheduled link checking should actually be run against the https://cfconventions.org/ site. That would cover the other repos that contribute to the web site, e.g., mailing-list-archive and cf-training.

are we scheduling (a.k.a cron) the checks and the resulting errors reported as an "whole" issue on the web repo? or when a PR against `main` it's been made as a auto-check? or both? [1]

[1] some old legacy links are already broken just because the page doesn't exist anymore.

**Comment by ethanrd** [comment]
Hi Antonio @cofinoa - Yes, I believe [this line](https://github.com/cf-convention/cf-convention.github.io/blob/main/.github/workflows/check_links.yml#L10) in the `check_links.yml` file causes the link checker to run once a week. A few lines above tell it to run for any PR against the "main" branch and below configures it so the link checker can be triggered manually.

It looks like each run includes the link checker running and writing a report and, I think, creating an issue. I don't see any "if failure then" for the creation of the issue. However, when the link checker runs against PR #298, it is not generating an issue instead giving an error `Error: Resource not accessible by integration` (e.g., [here](https://github.com/cf-convention/cf-convention.github.io/actions/runs/3155837203/jobs/5134925245)). I don't think we want issues created for PR runs but it looks like it tries to create an issue, I'm just not sure why it fails.

Regarding your footnote:
I fixed/updated/removed a number of old broken links in the last commit or two to PR #298. The last check against the PR only had 8 failing links (described above in my last comment).

**Comment by ethanrd** [comment]
Hi all - Just thought I would try to summarize the status of this issue. I think there have been about five topics under discussion:

1. There are lots of broken links. I believe PR #298 fixes a large number of the existing broken links, I think that PR is ready to be merged. There will still be some broken links which I think mainly fit into the categories summarized in my comment [above](https://github.com/cf-convention/cf-convention.github.io/issues/285#issuecomment-1263026462).
2. The link checker currently tries to create a new issue on completion (though it doesn't always succeed on runs against PRs). @DocOtak suggests [above](https://github.com/cf-convention/cf-convention.github.io/issues/285#issuecomment-1246970652) that instead it should add a comment to that PR. Another option is for it to do neither and simply fail with a non-zero exit status which would show up as a failed check in the PR. (I believe it does this now. Though the scheduled runs don't show up as failing which I don't understan.) 
3. @cofinoa suggests [above](https://github.com/cf-convention/cf-convention.github.io/issues/285#issuecomment-1259936636) that we use Liquid filters for referencing internal pages. I'm not familiar with liquid so unsure of the advantages over using absolute URLs (that start with a "/") or relative URLs (that don't start with a "/"). Also, not sure if they help with cross-repo references.
4. @cofinoa, as mentioned [here](https://github.com/cf-convention/cf-convention.github.io/issues/285#issuecomment-1261320776), has been running the link checker over the resulting HTML files in the `_site` directory rather than over the repo contents. If there are enough advantages, we could consider switching to that configuration. One downside is that it would create a dependence on the build action.
5. I suggested [above](https://github.com/cf-convention/cf-convention.github.io/issues/285#issuecomment-1261383535) that we consider changing the scheduled link checker to run against the main website (cfconventions.org) rather than against the repo contents as it does currently. That way it would check links in the full cfconventions.org site rather than just this repo.

Do those look good, did I miss any?

I think it would clean things up a bit to break these out into their own issues. I can work on that in a few days if it sounds good to everyone.

@erget - Thoughts on how this relates to issue #181? Perhaps I'll just link the resulting issues back to that one.

**Comment by erget** [comment]
@ethanrd thanks for working on this and summarising the points and devoting your attention to the issue!. I don't have an opinion on all of them but piece for piece:

1 - Thank you, I'm reviewing that right now.
2 - IIRC this is because you sent the PR from your fork, which is fine, but as a security feature GH doesn't let PRs initiated from forks spawn issues in the target repo. This is in line with the principle that actions inside a repo should be pulled in actively or initiated from the repo itself, so it's probably a good idea. **I expect that, if this is correct, the merged #298 will generate an issue due to the remaining bogus links.**
3 - No clue here.
4 - I don't see an issue with dependencies on different build actions, they should just be explicit, i.e. executing sequentially. I believe we can resolve this explicitly with [workflow_run](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_run). This separates concerns but still allows the build steps to interact, which I think is totally fine; in other pipelines you would test only after building, and then integrate only after that.
5 - this might make things simpler, I have no objections here. When experimenting with this a few years ago I tried that out and it was a lot slower, but I think at the time I was working with a different action with a different implementation, so that may no longer be a concern.

WRT #181 - I think this one is now outdated, as the link checker is working, just not like a well-oiled machine yet! I've closed it.

**Comment by erget** [comment]
Merged #298 - re-opening this one so we can continue following up but I think it's great to have those edits live now. Thanks @ethanrd !

**Comment by ethanrd** [comment]
Hi @erget - A few responses:

2 - Right! I should have thought of that. I've run into "issues" with PRs from forks before.
4 - I agree a dependency between build actions isn't a problem in general. However, I think such a dependency does add some complexity and so a solution that doesn't have that dependence should be preferred, all other things being equal.

I probably won't get back to this till Thursday and will look at splitting all these items into separate issues then.

**Comment by JonathanGregory** [comment]
Dear all

I see that a new "link checker report" issue is automatically generated each week. Does each one supersede the previous ones? If so, could the report instead be written to a file? Otherwise we will have to close these automatic issues regularly by hand. Can we do that now with the last four?

I see that most of the missing links are to (1) The [CEDA vocabulary editor](http://cfeditor.ceda.ac.uk/proposals) - has the URL changed, or is the vocab editor not working, do you know @japamment or @feggleton? (2) The UCAR email archive. I will check those (when I have time) and perhaps change them to the GitHub copy.

Jonathan

**Comment by feggleton** [comment]
Thanks @JonathanGregory Looks like that link needs a 1 at the end - https://cfeditor.ceda.ac.uk/proposals/1

**Comment by ethanrd** [comment]
Hi @JonathanGregory - I believe the CEDA vocabulary editor links are timing out rather than failing outright. I believe the link checker software has some settings for extending the length of the timeout that we might want to try. The remaining failing UCAR email archive links are, I believe, old links from before a long ago reorg of the archive. If I'm right, they will likely be difficult to fix as we'll have to look at the context where they are referenced to try and figure out which message they intend to reference.

Yes, each weekly link checker issue supersedes the previous ones. However, rather than change how we report the failure of the link checker test, I think we want to fix all currently broken links so that the test doesn't fail. That way new issues would only be created if an already referenced URL breaks and needs fixing.

**Comment by erget** [comment]
Hi folks,
I agree with @ethanrd 's assessment that it makes sense to cleanup the links first and then to leave the link checker running as-is. Would love to get to this myself but don't see it happening this month from my end.

WRT @JonathanGregory 's proposal to close the stale ones, I'm in favour and think that's fine.

Not sure about how to change the timeout settings, that is likely something we can tackle together with the overhaul of the links in general. We might also be able to make certain ones as `wontfix` or similar, although having misleading links doesn't really benefit anybody so I think it would make more sense to fix them or move on.

My 0.02€, hopefully I can contribute more actively again in the near future.

**Comment by JonathanGregory** [comment]
I'm going to close this now, because we have fixed or worked round the failing links.
